### PR TITLE
Stfc user office project 200 generic templates

### PR DESCRIPTION
**Pull request by vyshnavi1605**
## Description

Closes: https://github.com/UserOfficeProject/stfc-user-office-project/issues/200

Note: When downloading a submitted proposal, responses submitted using the generic template feature does not appear in pdf correctly, issue opened for this here: https://github.com/UserOfficeProject/stfc-user-office-project/issues/222

Replicated Sample Declaration functionality, in the form of Generic Templates. When logged in as a user officer, they can create as many templates as they want, and each are independent from each other. For example, using this feature they can create a `Grants` template and also `Publications` template and many more templates which they want to create.  

When logged in as a user, the user can fill out their answers to the generic templates created. The added feature allows them to input as many responses as they want to add (within the limits set by the user officer) using the `Add` button. The user can `Clone` and `Delete` the added response.

image
What user sees when submitting a proposal.

image
What user sees after clicking the `Add` button to add a `Grant`

image
When logged in as a user officer, they can create new templates and edit existing ones

image
Selecting `Add Sub Template` will allow a created generic template to be added into the questionnaire
## Motivation and Context

Allows customised questions to be asked in a straight forward process. Also, allows user to add, clone, delete their responses freely. 
https://github.com/UserOfficeProject/stfc-user-office-project/issues/200

## How Has This Been Tested

- Browser testing
   - Checked if process implemented was flowing correctly.
   - Logged in as a user officer, navigated to `Sub Templates` and created a template and added it to the proposal questionnaire.
   - Logged in as a user and filled out the questionnaire to check if everything was working as expected. 

- Frontend tests
   - Wrote frontend tests and ran them to see if they passed, if they didn't went through and made appropriate change to the code to ensure the tests passed.

## Fixes

Fixes: https://github.com/UserOfficeProject/stfc-user-office-project/issues/200

## Changes

- Added `GenericTemplate` folder in `components/questionary/questionaryComponents/GenericTemplate` which includes files that render and allow the generic template to be added to the proposal
- Added `GenericTemplateBasis` folder in `components/questionary/questionaryComponents/GenericTemplateBasis` which includes files that render and allows a generic template basis question to be added to each generic template created
- Added `genericTemplate` folder in `components/questionary/questionaries/genericTemplate`
- Modified `DashBoard.ts` and` MenuItems.ts`
- Added cypress tests in `generic_templates.ts`
- Modified `ErrorMessage.ts` to use error colours
- Added references to genericTemplates in `QuestionaryRegistry.tsx, QuestionaryDetails.tsx, QuestionaryComponentRegistry.tsx, ProposalQuestionaryDetails.tsx and ProposalCreate.tsx `
- Added `GraphQL` files for genericTemplates 
- Regenerated `SDK`

## Depends on

Backend PR: https://github.com/UserOfficeProject/user-office-backend/pull/425

## Tests included/Docs Updated?

- [ x ] I have added tests to cover my changes.
- [ x ] All relevant doc has been updated

### CODE REVIEW

File: src/components/questionary/questionaries/genericTemplate/GenericTemplateQuestionaryWizardStep.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import { QuestionaryWizardStep } from 'components/questionary/DefaultWizardStepFactory';
+
+export class GenericTemplateQuestionaryWizardStep extends QuestionaryWizardStep {
+  isItemWithQuestionaryEditable(): boolean {
+    return true; // Since view is locked when proposal is submitted, we can simply return true because edit sample will be disabled
```
Comment (TickleThePanda) [inline src/components/questionary/questionaries/genericTemplate/GenericTemplateQuestionaryWizardStep.ts:5]: Comment needs updating

File: src/hooks/genericTemplate/useDownloadPDFGenericTemplate.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+import {
+  DownloadContext,
+  PREPARE_DOWNLOAD_TYPE,
+} from 'context/DownloadContextProvider';
+
+export function useDownloadPDFGenericTemplate() {
+  const { prepareDownload } = useContext(DownloadContext);
+  const downloadGenericTemplatePDF = useCallback(
+    (sampleIds: number[], name: string) => {
```
Comment (TickleThePanda) [inline src/hooks/genericTemplate/useDownloadPDFGenericTemplate.ts:11]: Should this be `sampleIds` and `PDF_SAMPLE` here?

File: src/components/questionary/questionaryComponents/QuestionTemplateRelationFormShell.tsx [hunk 3]
```diff
@@ -108,7 +108,7 @@ export const QuestionTemplateRelationFormShell = (
         }}
         className={classes.naturalKey}
       >
-        {props.questionRel.question.naturalKey}
+        Click here to edit {props.questionRel.question.naturalKey}
```
Comment (TickleThePanda) [inline src/components/questionary/questionaryComponents/QuestionTemplateRelationFormShell.tsx:111]: I think `Edit {props.questionRel.question.naturalKey}` would be better.

File: src/graphql/template/fragment.fieldConfig.graphql [hunk 4]
```diff
@@ -109,4 +109,9 @@ fragment fieldConfig on FieldConfig {
     required
     tooltip
   }
+
+    ... on GenericTemplateBasisConfig {
```
Comment (TickleThePanda) [inline src/graphql/template/fragment.fieldConfig.graphql:113]: Whitespace looks wrong here.

File: cypress/types/template.d.ts [hunk 5]
```diff
@@ -114,6 +114,21 @@ declare global {
         options?: { minEntries?: number; maxEntries?: number }
       ) => void;
 
+      /**
+       * Creates generic tempale question
```
Comment (jekabs-karklins) [inline cypress/types/template.d.ts:118]: There is typo in "template"

**Pull request merged**